gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+        else:
+            # can put in a temporary tensor here
+            # mono['pe_result'] = pe_results[tensors] = (Indexed(Variable('PE'+str(len(pe_results)), (J,K)), (j,k)),)
+            # sometimes not all argument indices are present in the contraction
+            all_ind = set([fi for t1 in tensors for fi in t1.free_indices])
+            result_ind = tuple([i for i in arg_ind_flat if i in all_ind])
+            array = contract(tensors, quad_ind, result_ind)
+            pe_tensor = Indexed(Literal(array, 'PE'+str(len(pe_results))), result_ind)
+            mono['pe_result'] = pe_results[tensors] = (pe_tensor, )  # this is a tuple
+        # construct list of factor lists after pre-evaluation
+    factor_lists = []
+    for mono in monos_pe:
+        factor_lists.append(list(mono['rest'] + mono['pe_result']))
+
+    node_pe = sumproduct_2_node(factorise(factor_lists, quad_ind, arg_ind_flat))
+    node_pe = reorder(node_pe, arg_ind_flat)
+    theta_pe = count_flop(node_pe)  # flop count for pre-evaluation method
+    return (node_pe, theta_pe)
+
+
+def find_optimal_factors(monos, arg_ind_flat):
+    gem_int = OrderedDict()  # Gem node -> int
+    int_gem = OrderedDict()  # int -> Gem node
+    counter = 0
+    for mono in monos:
+        for j in arg_ind_flat:
+            # really this should just have 1 element
+            for n in mono[j]:
+                if n not in gem_int:
+                    gem_int[n] = counter
+                    int_gem[counter] = n
+                    counter += 1
+    if counter == 0:
+        return tuple()
+    # add connections (list of tuples)
+    edges_sets_list = []
+    # num_edges = dict.fromkeys(int_gem.keys(), 0)
+    for mono in monos:
+        # this double loop should be optimised further
+        edges_sets_list.append(tuple(
+            [gem_int[n] for j in arg_ind_flat for n in mono[j]]))
+
+    # product of extents of argument indices of nodes
+    extent = dict().fromkeys(int_gem.keys())
+    for key in extent:
+        arg_ind = set(int_gem[key].free_indices) & set(arg_ind_flat)
+        extent[key] = numpy.product([i.extent for i in arg_ind])
+    # set up the ILP
+    import pulp as ilp
+    prob = ilp.LpProblem('factorise', ilp.LpMinimize)
+    nodes = ilp.LpVariable.dicts('node', int_gem.keys(), 0, 1, ilp.LpBinary)
+
+    # objective function
+    big = 1000000  # some arbitrary big number
+    prob += ilp.lpSum(nodes[i]*(big - extent[i]) for i in int_gem)
+
+    # constraints (need to account for >2 argument indices)
+    for edges_set in edges_sets_list:
+        prob += ilp.lpSum(nodes[i] for i in edges_set) >= 1
+
+    prob.solve()
+    if prob.status != 1:
+        raise AssertionError("Something bad happened during ILP")
+
+    nodes_to_pull = tuple([n for n, node_number in gem_int.items()
+                           if nodes[node_number].value() == 1])
+    return nodes_to_pull
+
+
+def cse(node, quad_ind, arg_ind_flat):
+    """
+    common subexpression elimination
```

Comment:
So what does this function do? How does it do more than what automatically happens by value equality and DAG-aware code generation?